_posts/2017/2017-03-10-creating-a-sales-analytics-app-with-ruby-on-rails-2.md

```diff
@@ -0,0 +1,84 @@
+---
+layout: post
+title: "Creating a Sales Analytics App with Ruby on Rails, part 2"
+date: "2017-03-10 07:00:00"
+image: blog-header/rails-train-path-straight.jpg
+categories: tech-stories
+authors: ["Unai M."]
+---
+
+A couple of days ago we started writing about how and why we started our Sales analytics app. Today we want to talk about some issues we encountered in our way and how we solved them.
+
+## Issues on the development process
+
+As I wrote earlier, D3 is a really powerful engine to do all sorts of graphs but is by no means simple.
+This and the fact that we (backend, frontend and design) had separated tasks for the graphs created issues whenever we needed to change anything.
+The nice thing is that we work really close together and we are used to help each other, so in the end it turned out to be a really enriching experience for the whole team.
+
+The data the we use to display on the graphs and tables through the app comes obviously from each merchants shop.
+Currently, the API has a rate limit of 6000 calls per hour and although at the very beginning this seemed more than enough, it turned out to be an issue we had to find a solution for.
+
+Even from the start, it was clear for us that we had to persist the data in our own database; not only because of having limits on retrieving data but also we wanted to give the merchant extra value.
+For this, we needed to do some heavy calculations with data, so we couldn't rely on external systems.
+
+Our first approach was to add all the products, customers and orders to our database.
+For this, we were using [sidekiq](http://sidekiq.org/) to send the syncing tasks to the background.
+Sidekiq is fast and really customizable, but even with all the flexibility on the world, if the data isn't there, it can't be used.
+
+At the time of writing this, it seems obvious that shops aren't made of two demo customers and three orders with one product each.
+But when we where starting the project, that was our test data and this answer to the synchronization problem seemed to be, at least, a not bad one.
```

Comment:
... this was our test data and this answer to the synchronization challenge  seemed to be, at least, not a bad one.